core/src/ofac_stage.rs

```diff
@@ -0,0 +1,561 @@
+            .name("ofac_stage".into())
+            .spawn(move || {
+                while let Ok(mut packets) = verified_receiver.recv() {
+                    packets.0.iter_mut().for_each(|packet_batch| {
+                        discard_ofac_packets(
+                            packet_batch,
+                            &ofac_addresses,
+                            &address_lookup_table_cache,
+                        );
+                    });
+
+                    if ofac_sender.send(packets).is_err() {
+                        warn!("ofac_sender send error, returning early");
+                        return;
+                    }
+                }
+            })
+            .unwrap()
+    }
+}
+
+/// Discards packets that mention any OFAC related transactions
+fn discard_ofac_packets(
+    packet_batch: &mut PacketBatch,
+    ofac_addresses: &HashSet<Pubkey>,
+    address_lookup_table_cache: &DashMap<Pubkey, AddressLookupTableAccount>,
+) {
+    for p in packet_batch.iter_mut().filter(|p| !p.meta.discard()) {
+        let tx: bincode::Result<VersionedTransaction> = p.deserialize_slice(..);
+        if let Ok(tx) = tx {
+            if is_tx_ofac_related(&tx, ofac_addresses, address_lookup_table_cache) {
+                p.meta.set_discard(true);
+            }
+        }
+    }
+}
+
+/// Returns true if transaction is ofac-related, false if not
+fn is_tx_ofac_related(
+    tx: &VersionedTransaction,
+    ofac_addresses: &HashSet<Pubkey>,
+    address_lookup_table_cache: &DashMap<Pubkey, AddressLookupTableAccount>,
+) -> bool {
+    is_ofac_address_in_static_keys(tx, ofac_addresses)
+        || is_ofac_address_in_lookup_table(tx, ofac_addresses, address_lookup_table_cache)
+}
+
+/// Returns true if an ofac address is in the static keys for an account
+fn is_ofac_address_in_static_keys(
+    tx: &VersionedTransaction,
+    ofac_addresses: &HashSet<Pubkey>,
+) -> bool {
+    tx.message
+        .static_account_keys()
+        .iter()
+        .any(|acc| ofac_addresses.contains(acc))
+}
+
+/// Returns true if an ofac address is in the dynamic keys (lookup table) for an account
+fn is_ofac_address_in_lookup_table(
+    tx: &VersionedTransaction,
+    ofac_addresses: &HashSet<Pubkey>,
+    address_lookup_table_cache: &DashMap<Pubkey, AddressLookupTableAccount>,
+) -> bool {
+    if let Some(lookup_tables) = tx.message.address_table_lookups() {
+        for table in lookup_tables {
+            if let Some(lookup_info) = address_lookup_table_cache.get(&table.account_key) {
+                for idx in table
+                    .writable_indexes
+                    .iter()
+                    .chain(table.readonly_indexes.iter())
+                {
+                    if let Some(account) = lookup_info.addresses.get(*idx as usize) {
+                        if ofac_addresses.contains(account) {
+                            return true;
+                        }
+                    }
+                }
+            }
```

Comment:
If the tx references a lookup table that's not in our cache should we do something?  Force refresh the cache?  Automatically mark tx for discard since it's referencing a non-existent lookup table?